### PR TITLE
Add "forwarding optimization from new-bridge to tarjan (DON'T MERGE)

### PR DESCRIPTION
**Pull request by xmcclure**
See second commit.

The tarjan version of forwarding has significantly more code complexity than the version in new-bridge, since work while scanning an object has to be deferred until we know we're not forwarding. I have given this its own PR because I'm very hesitant to increase the code complexity by so much when testing does not show unambiguous performance improvements.

### CODE REVIEW

File: mono/metadata/sgen-tarjan-bridge.c [hunk 1]
```diff
@@ -517,6 +551,10 @@ push_object (GCObject *obj)
 #if DUMP_GRAPH
 		printf ("already marked\n");
 #endif
+
+		// Notice we only follow forward for non-initial objects (since forwarding is added on first scan)
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: What's a non-initial object?

File: mono/metadata/sgen-tarjan-bridge.c [hunk 2]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
+// However, when using OPTIMIZATION_FORWARD, we want to avoid pushing the object
+// if it has exactly one forwardable reference. Therefore we wait to push the
+// scan target until we know this criteria has failed.
+// initial_scan_one returns as arguments a forwardable argument (if any) and
+// a pushable object (if any). pushable objects are always forwardable.
+static void
+initial_scan_one (ScanData *data, GCObject *obj, ScanData **first_push, ScanData **first_forward, gboolean *forward_check_finished)
+{
+	ScanData *to_push = NULL, *to_forward = NULL;
+	initial_scan_check(data, obj, &to_push, &to_forward);
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: code formating.

File: mono/metadata/sgen-tarjan-bridge.c (hunk 1 above)
Comment (xmcclure) [inline mono/metadata/sgen-tarjan-bridge.c]: initial objects as in objects in state INITIAL. Would it be clearer if INITIAL were in all caps?

File: mono/metadata/sgen-tarjan-bridge.c [hunk 3]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
 // dfs () function's queue-children-of-object operation.
 static void
-push_all (ScanData *data)
+initial_scan (ScanData *data)
 {
 	GCObject *obj = data->obj;
 	char *start = (char*)obj;
 	mword desc = sgen_obj_get_descriptor_safe (obj);
 
+	gboolean forward_check_finished = FALSE;
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: Could you document how forwarding works in terms of forward_check_finished, first_push and first_forward?

The code is very hard to make sense of.

File: mono/metadata/sgen-tarjan-bridge.c [hunk 4]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
+{
+	data->low_index = data->index = object_index++;
 	dyn_array_ptr_push (&scan_stack, data);
+	dyn_array_ptr_push (&loop_stack, data);
+}
+
+// The target of the scan gets pushed onto the stack first.
+// However, when using OPTIMIZATION_FORWARD, we want to avoid pushing the object
+// if it has exactly one forwardable reference. Therefore we wait to push the
+// scan target until we know this criteria has failed.
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: Is scan target the "obj" variable? Then maybe name the variable that and use @scan_target in the comment to make the connection between docs and code more direct.

File: mono/metadata/sgen-tarjan-bridge.c [hunk 5]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
+	 * way by setting the forward pointer, and instead of
+	 * continuing processing this object, we start over with the
+	 * object it points to.
+	 */
+#ifdef OPTIMIZATION_FORWARD
+	if (!forward_check_finished) {
+		if (first_forward) {
+			ScanData *true_forward = follow_forward(first_forward);
+			// Don't allow loops. If this is a self-reference, pretend it isn't even there.
+			// Worth investigating: new-bridge didn't require this for OPTIMIZATION_FORWARD. Why does tarjan-bridge?
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: Perf issue or correctness issue?

File: mono/metadata/sgen-tarjan-bridge.c [hunk 6]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
 		GCObject *dst = (GCObject*)*(ptr);			\
-		if (dst) push_object (dst); 			\
+		if (dst)                                    \
+			initial_scan_one (data, dst, &first_push, &first_forward, &forward_check_finished ); \
 	} while (0)
 
 // dfs () function's queue-children-of-object operation.
 static void
-push_all (ScanData *data)
+initial_scan (ScanData *data)
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: Why rename the function? The new name doesn't convey any more meaning than the old one.
Plus, there's a comment mentioning the old function name.

File: mono/metadata/sgen-tarjan-bridge.c [hunk 7]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
+}
+
+// The target of the scan gets pushed onto the stack first.
+// However, when using OPTIMIZATION_FORWARD, we want to avoid pushing the object
+// if it has exactly one forwardable reference. Therefore we wait to push the
+// scan target until we know this criteria has failed.
+// initial_scan_one returns as arguments a forwardable argument (if any) and
+// a pushable object (if any). pushable objects are always forwardable.
+static void
+initial_scan_one (ScanData *data, GCObject *obj, ScanData **first_push, ScanData **first_forward, gboolean *forward_check_finished)
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: This function does not scanning, is initial_scan_one a good name?

File: mono/metadata/sgen-tarjan-bridge.c [hunk 8]
```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
...
+mark_object_scanned (ScanData *data)
+{
+	data->low_index = data->index = object_index++;
 	dyn_array_ptr_push (&scan_stack, data);
+	dyn_array_ptr_push (&loop_stack, data);
+}
+
+// The target of the scan gets pushed onto the stack first.
+// However, when using OPTIMIZATION_FORWARD, we want to avoid pushing the object
+// if it has exactly one forwardable reference. Therefore we wait to push the
```
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: I can't see how delaying the push changes anything under forwarding or non-forwarding.

**Comment by kumpera** [comment]
Can't we always push the forwarded to object in push_object/initial_scan_check? That would greatly simplify the code.

File: mono/metadata/sgen-tarjan-bridge.c (hunk 1 above)
Comment (kumpera) [inline mono/metadata/sgen-tarjan-bridge.c]: up to you.

**Comment by xmcclure** [comment]
@kumpera: I was curious about that, because there's so much state being slung around however I was not certain it would result in the same behavior. I'll re-investigate it.

**Comment by xmcclure** [comment]
I don't think we want this at all. I think that the existing "color" collapsing provides all the perf benefits this feature otherwise would, and does so without making the code unreadable.

**Pull request closed without merging**